python/basedetection.py

```diff
@@ -0,0 +1,319 @@
+        :return: True if initial analysis completed with results, False otherwise
+        :rtype: bool
+        """
+
+        if not arch and self._view_arch:
+            arch = str(self._view_arch)
+
+        if analysis not in ["basic", "controlFlow", "full"]:
+            raise ValueError("invalid analysis setting")
+
+        if alignment <= 0:
+            raise ValueError("alignment must be greater than 0")
+
+        if max_pointers < 2:
+            raise ValueError("max pointers must be at least 2")
+
+        if high_boundary < low_boundary:
+            raise ValueError("upper boundary must be greater than lower boundary")
+
+        settings = core.BNBaseAddressDetectionSettings(
+            arch.encode(),
+            analysis.encode(),
+            min_strlen,
+            alignment,
+            low_boundary,
+            high_boundary,
+            poi_analysis,
+            max_pointers,
+        )
+
+        if not core.BNDetectBaseAddress(self._handle, settings):
+            return False
+
+        max_candidates = 10
+        scores = (core.BNBaseAddressDetectionScore * max_candidates)()
+        confidence = core.BaseAddressDetectionConfidenceEnum()
+        last_base = ctypes.c_ulonglong()
+        num_candidates = core.BNGetBaseAddressDetectionScores(
+            self._handle, scores, max_candidates, ctypes.byref(confidence), ctypes.byref(last_base)
+        )
+
+        if num_candidates == 0:
+            return False
+
+        self._scores.clear()
+        for i in range(num_candidates):
+            self._scores.append((scores[i].BaseAddress, scores[i].Score))
+
+        self._confidence = confidence.value
+        self._last_tested_base_address = last_base.value
+        return True
+
+    def abort(self) -> None:
+        """
+        ``abort`` aborts base address detection analysis
+
+        .. note:: ``abort`` does not stop base address detection until after initial analysis has completed and it is \
+        in the base address enumeration phase
+
+        :rtype: None
+        """
+
+        core.BNAbortBaseAddressDetection(self._handle)
+
+    def get_reasons(self, base_address: int) -> list[BaseAddressDetectionReason]:
+        """
+        ``get_reasons`` returns a list of reasons that can be used to determine why a base address is a candidate
+
+        :param int base_address: base address to get reasons for
+        :return: list of reasons for the specified base address
+        :rtype: list[BaseAddressDetectionReason]
+        """
+
+        count = ctypes.c_size_t()
+        reasons = core.BNGetBaseAddressDetectionReasons(self._handle, base_address, ctypes.byref(count))
+        if count.value == 0:
+            return []
+
+        result = list()
+        for i in range(count.value):
+            result.append(BaseAddressDetectionReason(reasons[i].Pointer, reasons[i].POIOffset, reasons[i].POIType))
+        core.BNFreeBaseAddressDetectionReasons(reasons)
```

Comment:
we typically put `Free` calls in a finally block just to be absolutely sure it gets freed. So you need a `try` scope around your `reasons` lifetime and the free in the finally block